connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -159,6 +163,13 @@ private Long ensureLeaderConfig(long maxOffset, WorkerCoordinator coordinator) {
         // Base set: The previous assignment of connectors-and-tasks is a standalone snapshot that
         // can be used to calculate derived sets
         log.debug("Previous assignments: {}", previousAssignment);
+        int lastCompletedGenerationId = coordinator.lastCompletedGenerationId();
+        if (previousGenerationId != lastCompletedGenerationId) {
+            log.debug("Emptying previous assignments due to generation mismatch between previous "
+                    + "generation ID {} and last completed generation ID {} since the last assignment: {}",
+                    previousGenerationId, lastCompletedGenerationId, previousAssignment);
```

Comment:
This is related to the leader's internal bookkeeping when it calculates a new assignment. It's not related to the tasks that a worker (even the leader) is actually running. 

Emptying/clearing the previous assignment might result in some tasks shuffling around, because the leader will calculate an assignment from scratch, but it doesn't affect running tasks. The new computed assignment will send assignment and/or revocations as needed based on a) what tasks each worker has reported running in this round and which tasks are configured in the config topic. Another way to say this is that the leader won't bother detecting lost tasks in this round. Every unassigned task will be treated as a new task. 

You are right on the log message not conveying that meaning exactly. How about: 
```
log.debug("Clearing the view of previous assignments due to generation mismatch between "
                    + "previous generation ID {} and last completed generation ID {}. ",
                    previousGenerationId, lastCompletedGenerationId);
log.debug("This can happen if the leader fails to sync the assignment within a " 
                    + "rebalancing round. The following view of previous assignments might be "
                    + "outdated and will be ignored by the leader in the current computation of " 
                    + "new assignments. Possibly outdated previous assignments: {}", previousAssignment);
```

Reply:
Also, given that the previous assignments are printed in `debug`, I think it makes sense to keep these log messages in debug as well.

Reply:
Sounds good, though it'd be better to have a single (long) log message to prevent them from being separated by other log messages from other threads.

Reply:
I agree. The length will be very similar anyways. 
I'm pushing a commit to address your comments.